src/transport/SecureSessionTable.cpp

```diff
@@ -99,32 +105,78 @@ SecureSession * SecureSessionTable::EvictAndAllocate(uint16_t localSessionId, Se
 
     auto cleanup = MakeDefer([this]() { mRunningEvictionLogic = false; });
 
+#if CONFIG_BUILD_FOR_HOST_UNIT_TEST
+    const size_t kMaxSessionTableSize = mMaxSessionTableSize;
+#else
+    const size_t kMaxSessionTableSize = CHIP_CONFIG_SECURE_SESSION_POOL_SIZE;
+#endif
+
     ChipLogProgress(SecureChannel, "Evicting a slot for session with LSID: %d, type: %u", localSessionId,
                     (uint8_t) secureSessionType);
 
-    VerifyOrDie(mEntries.Allocated() <= CHIP_CONFIG_SECURE_SESSION_POOL_SIZE);
+    VerifyOrDie(mEntries.Allocated() <= kMaxSessionTableSize);
 
     //
     // Create a temporary list of objects each of which points to a session in the existing
     // session table, but are swappable. This allows them to then be used with a sorting algorithm
     // without affecting the sessions in the table itself.
     //
-    Platform::ScopedMemoryBufferWithSize<SortableSession> sortableSessions;
-    sortableSessions.Calloc(mEntries.Allocated());
-    if (!sortableSessions)
-    {
-        VerifyOrDieWithMsg(false, SecureChannel, "We couldn't allocate a session!");
-        return nullptr;
-    }
+    // The size of this shouldn't place significant demands on the stack - each item is
+    // 8 bytes in size (on a 32-bit platform), and 16 bytes in size (on a 64-bit platform,
+    // including padding). A minimal, spec compliant value for CHIP_CONFIG_SECURE_SESSION_POOL_SIZE
+    // is:
+    //
+    // (3 sessions / fabric * 5 fabrics + 1 reserved for CASE + 1 reserved for PASE) = 17 sessions.
+    //
+    // Total size of this stack variable = 17 * 8 = 136bytes (32-bit platform), 272 bytes (64-bit platform).
+    //
+    // We need to sort (as opposed to just a linear search for the smallest/largest item)
+    // since it is possible that the candidate selected for eviction may not actually be
+    // released once marked for expiration (see comments below for more details).
+    //
+    // Consequently, we may need to walk the candidate list till we find one that is.
+    // Sorting provides a better overall performance model in this scheme.
+    //
+    // (#19967): Investigate doing linear search instead.
+    //
+    //
+    SortableSession sortableSessions[CHIP_CONFIG_SECURE_SESSION_POOL_SIZE];
+
+    unsigned int index = 0;
+
+    //
+    // Compute two key stats for each session - the number of other sessions that
+    // match its fabric, as well as the number of other sessions that match its peer.
+    //
+    // This will be used by the session eviction algorithm later.
+    //
+    ForEachSession([&index, &sortableSessions, this](auto session) {
```

Comment:
```suggestion
    ForEachSession([&index, &sortableSessions, this](auto * session) {
```
To make it clear we are getting pointers, for purposes of the `!=` compare below.